state -> context api

dhrod0325 opened the pull request:
#2 
state 로 작성했던 코드를 context api로 변경 해 보았습니다

dhrod0325 committed:
refactor : jsx->js 확장자 변경

dhrod0325 committed:
bug : todoStorage location change

dhrod0325 committed:
refactor : Context Api로 교체

dhrod0325 committed:
refactor : useTodoService.js 메소드 네이밍 변경

dhrod0325 committed:
refactor : 함수 변경

parkgun6 commented on react-todo-props/src/components/TodoInput.jsx:
```diff
@@ -1,6 +1,9 @@
-export const TodoInput = ({onSubmitInput}) => {
     const [inputText, setInputText] = useState('');
 
     const inputRef = createRef();
```
createRef()를 사용하시는 이유가 있으실까요 ?
클래스형 컴포넌트가 아닌 함수형 컴포넌트에서의 useRef(), createRef()의 차이점과 createRef()를 했을때의 단점을 알아두시면 좋을것 같네요.

Ref에 대한 기본적인 이해는 하고있으시다고 생각하고, 참고하실만한 블로그 글 남겨드립니다!
https://kyounghwan01.github.io/blog/React/useRef-createRef/
useRef에 관한 공식 doc입니다.
https://ko.reactjs.org/docs/hooks-reference.html#useref

dhrod0325 replied:
좋은 정보 감사합니다! class로 컴포넌트를 작성할때 그냥 createRef를 사용해서 별 생각 없이 사용했는데 이런 차이가 있네요! 수정했습니다~!

parkgun6 commented on react-todo-props/src/components/TodoInput.jsx:
```diff
@@ -22,7 +25,7 @@ export const TodoInput = ({onSubmitInput}) => {
             return alert('할일을 입력하세요');
         }
 
-        onSubmitInput({message: inputText});
+        saveTodo({message: inputText});
```
saveTodo에 message props를 전달하는데에 message라는 변수명이 적절한지 의문을 가져보시는것도 좋으실 것 같습니다!
제 생각에는 '할 일을 입력하세요' 가 조금더 message에 가깝지 않나 라는 생각을 해봅니다.
그리고 alert에 들어가는 메세지는 constants 폴더를 만들어서
상수 메세지를 관리하는 js를 만들어보시는건 어떠실까요 ?
한 곳에서 메세지 관리를 해주시는게 관리차원에서 조금 더 유용하지 않을까 싶습니다!

dhrod0325 replied:
상수로 빼도록 해야겠어요!

parkgun6 commented on react-todo-props/src/components/TodoList.jsx:
```diff
@@ -1,37 +1,16 @@
-        };
-    }
+export const TodoList = () => {
+    const {todoList} = useTodoContext();
```
소스의 길이가 줄어든것이 보기좋습니다!
다만 가독성적인 면에서 
handleChangeCheck, handleClickRemove 등 이벤트를 관리하는 커스텀 훅의 이름이 useTodoContext라는 명명이 어울리는지 한번 더 고민해보시면 좋을것 같습니다!

dhrod0325 replied:
변경해 보았습니다!

parkgun6 commented on react-todo-props/src/components/TodoList.jsx:
```diff
@@ -1,37 +1,16 @@
-                            onSubmitModify={handleSubmitModify(todo)}
-                            onChangeCheck={handleChangeCheck(todo)}
-                            onClickRemove={handleClickRemove(todo)}
+                            todo={todo}
```
소스가 가벼워진것에 대해 아주 칭찬드리고싶습니다!

parkgun6 commented on react-todo-props/src/components/TodoModify.jsx:
```diff
@@ -16,13 +19,13 @@ export const TodoModify = ({text, onSubmitModify}) => {
 
-        onSubmitModify(modifyInput);
-
+        if (!modifyInput) return alert('할 일을 입력하세요');
+        updateTodo(todo, {message: modifyInput});
```
여기도 위에 남겨놓은 리뷰와 같은 맥락입니다!

parkgun6 commented on react-todo-props/src/hooks/useTodoService.js:
```diff
@@ -0,0 +1,49 @@
+export class TodoStorage {
+    STORAGE_KEY = 'todoList';
+
+    getList() {
+        const result = JSON.parse(localStorage.getItem(this.STORAGE_KEY));
+        return result ? result : [];
+    }
+
+    save(todoList) {
+        localStorage.setItem(this.STORAGE_KEY, JSON.stringify(todoList));
```
localStorage를 사용하여 Context API를 사용한 것에 대한 고민과 해결이 느껴집니다!
고생하셨습니다!
다만 useTodoService와 useTodoList는 커스텀훅이 아닌  context폴더를 구분해서 사용하시는것이 어떠실까요 ?

parkgun6 commented on react-todo-props/src/hooks/useTodoService.js:
```diff
@@ -0,0 +1,49 @@
+        save([...todoList]);
+    }, [todoList]);
+
+    function saveTodo(todo) {
+        const addTodo = {id: uuid(), checked: false, editState: false, ...todo};
+        setTodoList([...todoList, addTodo]);
+    }
+
+    function removeTodo(todo) {
+        todoList.splice(todoList.indexOf(todo), 1);
+        setTodoList([...todoList]);
+    }
+
+    function updateTodo(old, newTodo) {
+        const result = [...todoList];
+        result.splice(result.indexOf(old), 1, {...old, ...newTodo});
+        setTodoList(result);
+    }
+
+    function getList() {
+        return todoStorage.getList();
+    }
+
+    function save(todoList) {
+        todoStorage.save(todoList);
+    }
```
이곳에 화살표함수가 아닌 함수선언식으로 사용하신 이유가 있으실까요 ?

dhrod0325 replied:
```js
const [todoList, setTodoList] = useState(getList());
```
이 부분을 함수 선언 최상단에 올리고 싶은데 표현식으로 선언하면 const로 선언한 저 부분을 아래로 내려야 해서 function으로 변경했어요.! 다른 방법이 있는지 궁금해요~!

parkgun6 replied:
```suggestion

export const useTodoService = () => {
  const STORAGE_KEY = 'todoList';

  const getList = () => {
     const result = JSON.parse(localStorage.getItem(STORAGE.STORAGE_KEY));
     return result;
   };

  const save = (todoList) => {
    localStorage.setItem(STORAGE.STORAGE_KEY, JSON.stringify(todoList));
  };

  const [todoList, setTodoList] = useState(getList());

  useEffect(() => {
    save([...todoList]);
  }, [todoList]);

  const saveTodo = (todo) => {
    const addTodo = { id: uuid(), checked: false, editState: false, ...todo };
    setTodoList([...todoList, addTodo]);
  };

  const removeTodo = (todo) => {
    todoList.splice(todoList.indexOf(todo), 1);
    setTodoList([...todoList]);
  };

  const updateTodo = (old, newTodo) => {
    const result = [...todoList];
    result.splice(result.indexOf(old), 1, { ...old, ...newTodo });
    setTodoList(result);
  };

  return { todoList, saveTodo, removeTodo, updateTodo };
};

```

parkgun6 replied:
TodoStorage를 커스텀훅에 class로 선언하신 이유가 있으실까요 ?
localStorage만 다루는 파일을 분리해서 재사용성에 대해서 고려해보시는것도 좋으실 것 같습니다.
서비스 로직의 위치도 고려하셔서 위치를 다시한번 고려해보시면 좋으실 것 같습니다.
커스텀 훅의 규칙에 대해서도 알아보시면 좋으실 것 같습니다!
고생하셨습니다.

parkgun6 replied:
```suggestion

export const TodoService = () => {
  const STORAGE_KEY = 'todoList';
  
  const getList = () => {
    return JSON.parse(localStorage.getItem(STORAGE_KEY));
  };

  const save = (todoList) => {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(todoList));
  };
  const [todoList, setTodoList] = useState(getList());

  useEffect(() => {
    save([...todoList]);
  }, [todoList]);

  const saveTodo = (todo) => {
    const addTodo = { id: uuid(), checked: false, editState: false, ...todo };
    setTodoList([...todoList, addTodo]);
  };

  const removeTodo = (todo) => {
    todoList.splice(todoList.indexOf(todo), 1);
    setTodoList([...todoList]);
  };

  const updateTodo = (old, newTodo) => {
    const result = [...todoList];
    result.splice(result.indexOf(old), 1, { ...old, ...newTodo });
    setTodoList(result);
  };

  return { todoList, saveTodo, removeTodo, updateTodo };
};

```
STORAGE_KEY 는 상수로 빼는게 좋지않을까요 ?
TodoStorage를 커스텀훅에 class로 선언하신 이유가 있으실까요 ?
getList와 save는 서비스로직으로 분리하고, 
localStorage만 다루는 파일을 분리해서 재사용성에 대해서 고려해보시는것도 좋으실 것 같습니다.
서비스 로직의 위치도 고려하셔서 위치를 다시한번 고려해보시면 좋으실 것 같습니다.
커스텀 훅의 규칙에 대해서도 알아보시면 좋으실 것 같습니다!
고생하셨습니다.

parkgun6 reviewed (commented):
커스텀훅에 대해서 찾아보고 고민과 해결을 하셨다는 고생이 느껴집니다!
고생많으셨다는 말을 먼저 드리고 싶습니다!
읽다보면 변수명이 무엇을 의미하는지 알쏭달쏭한 부분들이 있어서
그 부분에 대해 리뷰 남겨놓았습니다!
Context API와 localStorage를 사용하신것에 대해 칭찬드리고 싶습니다.
다만 Context는 context폴더를 구분하여 사용하시는 것은 어떠실까요 ?
Props로 내린것보다 상태관리를 통하여 소스가 많이 줄어들었음에 가독성이 많이 올라갔다고 느껴집니다.
고생 많으셨습니다!

dhrod0325 committed:
refactor : 스터디별 디렉토리 변경중
recoil,context,react-query를 디렉토리 별로 분리하여 공부해 보기 위함
LocalStorage.js, TodoService.js utils 폴더로 이동

dhrod0325 committed:
style : 리뷰 반영(네이밍)
메소드명, 변수명 수정

dhrod0325 committed:
refactor : 잘못 사용하던 코드 수정
useEffect에서 ref값을 직접 변경하던것올 input에 value로 넘김

dhrod0325 committed:
style : 네이밍 변경

parkgun6 commented on react-todo-props/src/hooks/useTodoContext.jsx:
```diff
@@ -0,0 +1,17 @@
+    const context = React.useContext(TodoContext);
+
+    if (context === undefined) {
+        throw new Error();
```
에러를 던지는 로직 좋습니다.

dhrod0325 committed:
refactor : 잘못된 의미의 hook 사용 제거

dhrod0325 committed:
refactor : service,storage 변경

dhrod0325 committed:
refactor : context api-> recoil 변경

dhrod0325 committed:
refactor : 폴더명 변경
